src/model/position.js

```diff
@@ -459,20 +476,34 @@ export default class Position {
 					return false;
 				}
 
-				left.path = left.path.slice( 0, -1 );
+				const path = left.getParentPath();
+				path[ path.length - 1 ]++;
+				left = new Position( left.root, path );
+
 				leftParent = leftParent.parent;
-				left.offset++;
 			} else {
 				if ( right.offset !== 0 ) {
 					return false;
 				}
 
-				right.path = right.path.slice( 0, -1 );
+				right = new Position( right.root, right.getParentPath() );
 			}
 		}
 	}
 
 	/**
+	 * Converts `Position` to plain object and returns it.
+	 *
+	 * @returns {Object} `Position` converted to plain object.
+	 */
+	toJSON() {
+		return {
+			root: this.root.toJSON(),
+			path: this.path
+		};
+	}
+
+	/**
 	 * Returns a copy of this position that is updated by removing `howMany` nodes starting from `deletePosition`.
```

Comment:
Docs might need an update if we won't be returning a copy in every scenario (see comment in code below).